src/RefractDataStructure.cc

```diff
@@ -1289,76 +1336,188 @@ namespace
+    std::unique_ptr<OptionElement> MsonOneOfEntryToElements(const mson::Element::PropertyMemberSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context)
+    {
+        if (auto apie = MsonTypeSectionToElements(section, sourceMap, context, mson::StringTypeName))
+            return make_element<OptionElement>(std::move(apie));
+        return nullptr;
+    }
 
-        for (const auto& oneOfInfo : oneOfNodeInfo) {
+    std::unique_ptr<OptionElement> MsonOneOfEntryToElements(const mson::Element::ValueMemberSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context)
+    {
+        if (auto apie = MsonTypeSectionToElements(section, sourceMap, context, mson::StringTypeName))
+            return make_element<OptionElement>(std::move(apie));
+        return nullptr;
+    }
 
-            auto option = make_element<OptionElement>();
+    std::unique_ptr<OptionElement> MsonOneOfEntryToElements(const mson::Element::MixinSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context)
+    {
+        if (auto apie = MsonTypeSectionToElements(section, sourceMap, context, mson::StringTypeName))
+            return make_element<OptionElement>(std::move(apie));
+        return nullptr;
+    }
 
-            // we can not use MsonElementToRefract() for groups,
-            // "option" element handles directly all elements in group
-            if (oneOfInfo.node->klass == mson::Element::GroupClass) {
-                MsonElementsToRefract(MakeNodeInfo(oneOfInfo.node->content.elements(), oneOfInfo.sourceMap->elements()),
-                    std::back_inserter(option->get()),
-                    context);
-            } else {
-                if (auto apie = MsonElementToRefract(oneOfInfo, context, mson::StringTypeName)) {
-                    option->get().push_back(std::move(apie));
-                }
-            }
+    std::unique_ptr<OptionElement> MsonOneOfEntryToElements(const mson::Element::OneOfSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context)
+    {
+        if (auto apie = MsonTypeSectionToElements(section, sourceMap, context, mson::StringTypeName))
+            return make_element<OptionElement>(std::move(apie));
+        return nullptr;
+    }
 
-            select->get().push_back(std::move(option));
+    std::unique_ptr<OptionElement> MsonOneOfEntryToElements(const mson::Element::GroupSection& section,
+        const snowcrash::SourceMap<mson::Element>* sourceMap,
+        ConversionContext& context)
+    {
+        auto option = make_element<OptionElement>();
+        MsonElementsToRefract( //
+            MakeNodeInfo(*section, sourceMap->elements()),
+            std::back_inserter(option->get()),
+            context);
+        return option;
+    }
+
+    struct MsonOneOfEntryToElementsLambda {
```

Comment:
This can be a much shorter lambda in C++14. In C++11 we don't have a choice but introducing a structure with a template operator().